addon/index.d.ts

```diff
@@ -0,0 +1,940 @@
+
+  /**
+   * Configures the task group to run task instances
+   * one-at-a-time in the order they were `.perform()`ed.
+   * Sets default maxConcurrency to 1.
+   *
+   * @method enqueue
+   * @memberof TaskGroupProperty
+   * @instance
+   */
+  enqueue(): this;
+
+  /**
+   * Configures the task group to immediately cancel (i.e.
+   * drop) any task instances performed when the task group
+   * is already running at maxConcurrency. Sets default
+   * maxConcurrency to 1.
+   *
+   * @method drop
+   * @memberof TaskGroupProperty
+   * @instance
+   */
+  drop(): this;
+
+  /**
+   * Configures the task group to drop all but the most
+   * recently performed {@linkcode TaskInstance }.
+   *
+   * @method keepLatest
+   * @memberof TaskGroupProperty
+   * @instance
+   */
+  keepLatest(): this;
+
+  /**
+   * Sets the maximum number of task instances that are
+   * allowed to run in this task group at the same time.
+   * By default, with no task modifiers applied, this number
+   * is Infinity (there is no limit to the number of tasks
+   * that can run at the same time).
+   * {@linkcode TaskGroupProperty#restartable .restartable()},
+   * {@linkcode TaskGroupProperty#enqueue .enqueue()}, and
+   * {@linkcode TaskGroupProperty#drop .drop()} set the
+   * default maxConcurrency to 1, but you can override this
+   * value to set the maximum number of concurrently running
+   * tasks to a number greater than 1.
+   *
+   * [See the AJAX Throttling example](/#/docs/examples/ajax-throttling)
+   *
+   * The example below uses a task group with `maxConcurrency(3)`
+   * to limit the number of concurrent AJAX requests (for anyone
+   * using tasks in this group) to 3.
+   *
+   * ```js
+   * ajax: taskGroup().maxConcurrency(3),
+   *
+   * doSomeAjax: task(function * (url) {
+   *   return Ember.$.getJSON(url).promise();
+   * }).group('ajax'),
+   *
+   * doSomeAjax: task(function * (url) {
+   *   return Ember.$.getJSON(url).promise();
+   * }).group('ajax'),
+   *
+   * elsewhere() {
+   *   this.get('doSomeAjax').perform("http://www.example.com/json");
+   * },
+   * ```
+   *
+   * @method maxConcurrency
+   * @memberof TaskGroupProperty
+   * @param {Number} n The maximum number of concurrently running tasks
+   * @instance
+   */
+  maxConcurrency(n: number): this;
+}
+
+export type TaskCancelation = Error & { name: 'TaskCancelation' };
+
+export type Yieldable<T = void> = PromiseLike<T>;
+
+type Evented = {
+  on(event: string, callback: (...args: any[]) => void): void;
+  off(event: string, callback: (...args: any[]) => void): void;
+} | {
+  one(event: string, callback: (...args: any[]) => void): void;
+} | {
+  addEventListener(event: string, callback: (...args: any[]) => void): void;
+  removeEventListener(event: string, callback: (...args: any[]) => void): void;
+};
+
+type Resolved<T> = T extends PromiseLike<infer R> ? R : T;
+
+type Settlement<T> = { state: 'fulfilled', value: T } | { state: 'rejected', reason: any };
```

Comment:
We discussed the tradeoffs around `any` vs. `unknown` elsewhere, but on this one I'm more inclined toward `unknown`, as it "forces" the right response for users, in a good way. Thoughts?

Reply:
I think I copied this from the built-in `Promise` type. Since `TaskInstance<T>` implements `Promise<T>`, I figured it's good to match the built-in promise things as close as possible to avoid any accidental divergances and assignability issues. The rest of it falls out from that decision.

Reply:
Ah, very good; that makes sense. I dislike it, but that’s been there for a *long* time and would break a lot of existing code, so I understand why they’re leery of changing it. Thanks!